Makefile

```diff
@@ -68,6 +75,12 @@ kernel-hifive-2:
 print-sizes: examples
 	cargo run --release -p print_sizes
 
+# Runs a libtock2 example in QEMU on a simulated HiFive board.
+.PHONY: qemu-example
+qemu-example: kernel-hifive-2
```

Comment:
This should use the platform specified by the user. QEMU supports multiple Tock platforms

Reply:
I'm going to draw a distinction between the `runner` binary and `make qemu-example`, and give two responses here.

## `runner`

The only other Tock platform with QEMU support I'm aware of is OpenTitan. I added support for OpenTitan to the runner, and added a `kernel-opentitan` `make` action to build a suitable kernel. It should be easy to add support for other platforms.

It successfully launches QEMU, but appears to hang after jumping into the kernel:

```
jrvanwhy@penguin:~/libtock-rs$ LIBTOCK_PLATFORM="opentitan" cargo run --release -p libtock2 --example low_level_debug --target=riscv32imc-unknown-none-elf -- --deploy qemu
    Finished release [optimized + debuginfo] target(s) in 0.66s
     Running `cargo run -p runner --release target/riscv32imc-unknown-none-elf/release/examples/low_level_debug --deploy qemu`
    Finished release [optimized + debuginfo] target(s) in 0.04s
     Running `target/release/runner target/riscv32imc-unknown-none-elf/release/examples/low_level_debug --deploy qemu`
I00000 boot_rom.c:52] Version:    opentitan-snapshot-20191101-1-3845-g000a18cb6
Build Date: 2021-01-11, 18:42:51

I00001 boot_rom.c:61] Boot ROM initialisation has completed, jump into flash!
qemu-system-riscv32: terminating on signal 2 from pid 1219 (target/release/runner)
```

## `make qemu-example`

I don't think that `qemu-example` should support multiple platforms:

1. It is meant to be a simple "run my example" command. Forcing users to specify `LIBTOCK_PLATFORM` every time makes it more time consuming to use, and I don't expect it will be useful the majority of the time.
2. We don't have a way to pass the correct `--target` parameter. We could just build for `riscv32-imc-unknown-none-elf`, because that will work for both HiFive1 and OpenTitan, but that's a hack.

Instead, I think users who want to test on a non-HiFive1 platform should invoke `cargo` directly.

Reply:
@alistair23 Are you happy with my response, or are there further changes you'd like?

Reply:
For the `runner`, it seems like there is a bug in this PR then, as it should run the kernel and the app

For `make qemu-example` I agree we should by default just run, but we should also respect the users `LIBTOCK_PLATFORM` if they supply one. Also, due to https://github.com/tock/libtock-rs/issues/156 I think we should use OT as the default instead of the HiFive1

Reply:
I was under the assumption the issue was in the Tock kernel. If there's a bug in this PR it's probably in the arguments to either `elf2tab` or QEMU. Here's the output of a run with `--verbose`, perhaps you can spot an error in the arguments?

```
jrvanwhy@penguin:~/libtock-rs$ LIBTOCK_PLATFORM="opentitan" cargo run --release -p libtock2 --example low_level_debug --target=riscv32imc-unknown-none-elf -- --deploy qemu --verbose
    Finished release [optimized + debuginfo] target(s) in 0.04s
     Running `cargo run -p runner --release target/riscv32imc-unknown-none-elf/release/examples/low_level_debug --deploy qemu --verbose`
    Finished release [optimized + debuginfo] target(s) in 0.04s
     Running `target/release/runner target/riscv32imc-unknown-none-elf/release/examples/low_level_debug --deploy qemu --verbose`
Package name: "low_level_debug"
TAB path: target/riscv32imc-unknown-none-elf/release/examples/low_level_debug.tab
Protected region size: 72
Found .stack section, size: 256
ELF file: "target/riscv32imc-unknown-none-elf/release/examples/low_level_debug"
TBF path: target/riscv32imc-unknown-none-elf/release/examples/low_level_debug.tbf
elf2tab command: "elf2tab" "--kernel-major" "2" "--kernel-minor" "0" "-n" "low_level_debug" "-o" "target/riscv32imc-unknown-none-elf/release/examples/low_level_debug.tab" "--protected-region-size" "72" "--stack" "256" "target/riscv32imc-unknown-none-elf/release/examples/low_level_debug" "-v"
Spawning elf2tab
Creating "target/riscv32imc-unknown-none-elf/release/examples/low_level_debug.tbf"
Min RAM size from segments in ELF: 0 bytes
Number of writeable flash regions: 0
Kernel version: 2.0
Entry point is in .start section
  Adding .start section. Offset: 72 (0x48). Length: 118 (0x76) bytes.
  Adding 2 bytes of padding between sections
  Adding .text section. Offset: 192 (0xc0). Length: 64 (0x40) bytes.
Searching for .rel.X sections to add.
TBF Header:
               version:        2        0x2
           header_size:       72       0x48
            total_size:      260      0x104
                 flags:        1        0x1

        init_fn_offset:       32       0x20
        protected_size:        0        0x0
      minimum_ram_size:     2304      0x900

     start_process_ram: 268451840 0x10004000
   start_process_flash: 537067592 0x20030048

    kernel version: ^2.0

elf2tab finished. exit status: 0
Detected platform opentitan
QEMU command: "tock2/tools/qemu/build/qemu-system-riscv32" "-device" "loader,file=target/riscv32imc-unknown-none-elf/release/examples/low_level_debug.tbf,addr=0x20030000" "-nographic" "-serial" "mon:stdio" "-bios" "tock2/tools/qemu-runner/opentitan-boot-rom.elf" "-kernel" "tock2/target/riscv32imc-unknown-none-elf/release/earlgrey-cw310" "-M" "opentitan"
Spawning QEMU
I00000 boot_rom.c:52] Version:    opentitan-snapshot-20191101-1-3845-g000a18cb6
Build Date: 2021-01-11, 18:42:51

I00001 boot_rom.c:61] Boot ROM initialisation has completed, jump into flash!
qemu-system-riscv32: terminating on signal 2 from pid 3077 (target/release/runner)
Waiting for child process.
```

Reply:
We run Tock on OT as part of the kernel CI, so it shouldn't be a kernel bug.

Are you sure the submodule is updated? Maybe try forcing a clean and reset

Reply:
> Are you sure the submodule is updated? Maybe try forcing a clean and reset

This was it, thanks! I tried to run the QEMU tests in the kernel tree, and QEMU didn't recognize the `-semihosting` flag, so I think the problem was that I forgot to rebuild QEMU after updating the submodule.

Now, it usually works:

```
jrvanwhy@penguin:~/libtock-rs$ LIBTOCK_PLATFORM="opentitan" cargo run --release -p libtock2 --example low_level_debug --target=riscv32imc-unknown-none-elf -- --deploy qemu
    Finished release [optimized + debuginfo] target(s) in 0.04s
     Running `cargo run -p runner --release target/riscv32imc-unknown-none-elf/release/examples/low_level_debug --deploy qemu`
    Finished release [optimized + debuginfo] target(s) in 0.04s
     Running `target/release/runner target/riscv32imc-unknown-none-elf/release/examples/low_level_debug --deploy qemu`
I00000 boot_rom.c:60] Version:    opentitan-earlgrey_silver_release_v5-157-ga28c280bb
Build Date: 2021-06-30, 02:13:50

I00001 boot_rom.c:76] Boot ROM initialisation has completed, jump into flash!
OpenTitan initialisation complete. Entering main loop
LowLevelDebug: App 0x0 prints 0x1
LowLevelDebug: App 0x0 prints 0x2 0x3
qemu-system-riscv32: terminating on signal 2 from pid 21752 (target/release/runner)
```

Sometimes though, it misses the final line of output:

```
jrvanwhy@penguin:~/libtock-rs$ LIBTOCK_PLATFORM="opentitan" cargo run --release -p libtock2 --example low_level_debug --target=riscv32imc-unknown-none-elf -- --deploy qemu
    Finished release [optimized + debuginfo] target(s) in 0.04s
     Running `cargo run -p runner --release target/riscv32imc-unknown-none-elf/release/examples/low_level_debug --deploy qemu`
    Finished release [optimized + debuginfo] target(s) in 0.04s
     Running `target/release/runner target/riscv32imc-unknown-none-elf/release/examples/low_level_debug --deploy qemu`
I00000 boot_rom.c:60] Version:    opentitan-earlgrey_silver_release_v5-157-ga28c280bb
Build Date: 2021-06-30, 02:13:50

I00001 boot_rom.c:76] Boot ROM initialisation has completed, jump into flash!
OpenTitan initialisation complete. Entering main loop
LowLevelDebug: App 0x0 prints 0x1
qemu-system-riscv32: terminating on signal 2 from pid 21725 (target/release/runner)
```

so *something* isn't working reliably. I don't see the same issue in HiFive1, so I'm inclined to leave `make qemu-example` on HiFive1. If we find and solve the issue, then I think we should switch `make qemu-example` to OpenTitan.

Does that sound good to you?

Reply:
Hmmm... That's strange. Are you sure the line never appears? Is it possible it's just getting killed by the runner too soon?

Reply:
I gave it a few seconds before I `Ctrl+C`'d it. I also tried typing a bit to see if that would make something wake up, but nothing happened.

Reply:
Hmm... Ok, well let's add this to future problems :)